### ISSUE TITLE
[#Wave-Control - Add Sage Intacct] Add connection sync progress and error handling.

### ISSUE THREAD

**Issue by hungvu193** [Expensify/App#44669]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

Tracking GH: https://github.com/Expensify/Expensify/issues/388780

Design doc section: 
https://docs.google.com/document/d/1k3ZFw8KB55yPUSCG6KYZlwpwEtmRt3eUshwxs7bZq5I/edit#heading=h.8m1ckdwmo1e5

A follow up of https://github.com/Expensify/App/issues/43532

<details><summary>Issue Owner</summary>Current Issue Owner: @miljakljajic</details>

**Comment by hungvu193** [comment]
cc @yuwenmemon @SzymczakJ

**Comment by SzymczakJ** [comment]
Hey! I’m Jakub Szymczak from Software Mansion, an expert agency, and I’d like to work on this issue!

**Comment by yuwenmemon** [comment]
Thanks @hungvu193 !

**Comment by SzymczakJ** [comment]
Couple questions on how thing on BE are going:
1. What about sending Sage Intacct sync progress from BE like we do for QBO/Xero?
2. Is `UpdateSageIntacctEntity` endpoint ready? When connecting to Sage Intacct for the first time, there's no `config.entity` field. Shouldn't we provide connection with some initial setting like we do on Xero with Xero Organization(one of organizations is chosen arbitrarily)?
3. Bumping my [question](https://swmansion.slack.com/archives/C06ML6X0W9L/p1719998266463689?thread_ts=1719271507.184899&cid=C06ML6X0W9L) that I made on Slack

**Comment by NikkiWines** [comment]
> What about sending Sage Intacct sync progress from BE like we do for QBO/Xero?

Pretty sure @yuwenmemon added this already (https://github.com/Expensify/Integration-Server/pull/7974, right, @yuwenmemon?) 

> Is UpdateSageIntacctEntity endpoint ready? When connecting to Sage Intacct for the first time, there's no config.entity field. Shouldn't we provide connection with some initial setting like we do on Xero with Xero Organization(one of organizations is chosen arbitrarily)?

Yep, the `UpdateSageIntacctEntity` endpoint is on prod already. If the user is using the `Top level` entity we store an empty string, otherwise we store the ID associated with the entity. I believe by default we use the `Top level` (if available) which might be why you're not seeing an initial value 

> Bumping my [question](https://swmansion.slack.com/archives/C06ML6X0W9L/p1719998266463689?thread_ts=1719271507.184899&cid=C06ML6X0W9L) that I made on Slack
cc @yuwenmemon @NikkiWines

Looking at this now, will answer in the thread

**Comment by NikkiWines** [comment]
> What about an endpoint for connecting by reusing other policy's credentials? I remember that at some point we talked about it, but it's not mentioned in doc.

I'm not sure about this! I'll take a look at what we do for Expensify Classic now.

> What about SyncPolicyToSageIntacct API command, is it already up?

Yes, this is already on prod!

**Comment by SzymczakJ** [comment]
Another batch of questions:
1. `SyncPolicyToSageIntacct` is working, but it's not sending sync progress, `connectionSyncProgress` is not updated by BE. The same happens when we connect to Sage Intacct for the first time. Is there any estimate on when this will be ready?
2. What about an endpoint for connecting by reusing other policy's credentials? Or maybe we're able to reuse some existing endpoint?
3. What about error handling in case when user gives wrong credentials? Right now any errors that happen with Sage Intacct sync are put into `policy.connections` object, but when user gives wrong credentials we're not going to connect him to Sage Intacct and therefore there will be no `policy.connections` object and no error. Related [slack thread].(https://swmansion.slack.com/archives/C06ML6X0W9L/p1720179157722519?thread_ts=1719271507.184899&cid=C06ML6X0W9L)
<img width="1047" alt="Screenshot 2024-07-05 at 13 13 56" src="https://github.com/Expensify/App/assets/88395093/088225a2-2ee1-4d67-a73f-cd338b16f7f2">
4. Found this comment related to tax sync in the design doc: 

> Tax will only be an option if the customer is syncing at an entity with tax groups set up in Sage Intacct OR if they are a multi-entity set up with tax groups set up, syncing at the top level.

https://docs.google.com/document/d/1k3ZFw8KB55yPUSCG6KYZlwpwEtmRt3eUshwxs7bZq5I/edit#bookmark=id.myrsjl26u0g7

How do I know that the entity has a tax group set up? Maybe we should put this in `connections.intacct.data.entities`?

Do you know answers to these questions? @JmillsExpensify @yuwenmemon @NikkiWines

**Comment by NikkiWines** [comment]
> SyncPolicyToSageIntacct is working, but it's not sending sync progress, connectionSyncProgress is not updated by BE. The same happens when we connect to Sage Intacct for the first time. Is there any estimate on when this will be ready?

@yuwenmemon can you check this out since you added some sync progress logic [here](https://github.com/Expensify/Integration-Server/pull/7974)?

> What about an endpoint for connecting by reusing other policy's credentials? Or maybe we're able to reuse some existing endpoint?

I think this is more of a polish than a core requirement, so let's make this a follow-up issue to complete. What do you think @yuwenmemon?

> What about error handling in case when user gives wrong credentials? Right now any errors that happen with Sage Intacct sync are put into policy.connections object, but when user gives wrong credentials we're not going to connect him to Sage Intacct and therefore there will be no policy.connections object and no error. Related [slack thread].(https://swmansion.slack.com/archives/C06ML6X0W9L/p1720179157722519?thread_ts=1719271507.184899&cid=C06ML6X0W9L)

It doesn't appear like we currently store anything related to the error when the sync fails - we just display the error on Expensify Classic and clear out the connection data. I can update this so we store the error in the `policy.connections` object 

> How do I know that the entity has a tax group set up? Maybe we should put this in connections.intacct.data.entities?

I think the best way to confirm this is to check that `data.entities` is not empty and also that `config.tax.syncTax` is present in the connection data.

**Comment by SzymczakJ** [comment]
> > What about error handling in case when user gives wrong credentials? Right now any errors that happen with Sage Intacct sync are put into policy.connections object, but when user gives wrong credentials we're not going to connect him to Sage Intacct and therefore there will be no policy.connections object and no error. Related [slack thread].(https://swmansion.slack.com/archives/C06ML6X0W9L/p1720179157722519?thread_ts=1719271507.184899&cid=C06ML6X0W9L)
> 
> It doesn't appear like we currently store anything related to the error when the sync fails - we just display the error on Expensify Classic and clear out the connection data. I can update this so we store the error in the policy.connections object

I ask because we clearly have such pattern in Sage Intacct design doc: 1. User wants to connect to Sage Intacct but gives wrong credentials. 2. We indicate that and allow user to Enter credentials again(to achieve these two things we need connections data)
<img width="737" alt="Screenshot 2024-07-08 at 15 40 59" src="https://github.com/Expensify/App/assets/88395093/9d669918-adb1-43a9-8702-12d144e7119d">

**Comment by yuwenmemon** [comment]
>I think this is more of a polish than a core requirement, so let's make this a follow-up issue to complete. What do you think @yuwenmemon?

I'm adding an endpoint to reuse credentials here: https://github.com/Expensify/App/issues/44718 - Feel free to follow the API outlined there.

**Comment by yuwenmemon** [comment]
> SyncPolicyToSageIntacct is working, but it's not sending sync progress, connectionSyncProgress is not updated by BE. The same happens when we connect to Sage Intacct for the first time. Is there any estimate on when this will be ready?

Yeah it _should_ be working but I'll look into why it might not be.

**Comment by yuwenmemon** [comment]
> It doesn't appear like we currently store anything related to the error when the sync fails - we just display the error on Expensify Classic and clear out the connection data. I can update this so we store the error in the policy.connections object

@NikkiWines Are you drafting a PR for this?

**Comment by SzymczakJ** [comment]
> I can update this so we store the error in the policy.connections object

How would we do it? My initial idea would be to add Sage Intacct to `policy.connections` object as soon as user calls `CONNECT_POLICY_TO_SAGE_INTACCT` API command and when password or userID turns out to be wrong, then set error in `policy.connections`. Of course `policy.connections` object should be populated with the rest of data only when the credentials are correct

**Comment by NikkiWines** [comment]
Yep! Got one in the works!

**Comment by NikkiWines** [comment]
PR(s) are in review for this: 

- https://github.com/Expensify/Integration-Server/pull/7987
- https://github.com/Expensify/Web-Expensify/pull/42692/

**Comment by NikkiWines** [comment]
Both PRs have been merged - we're waiting on the web-e PR to be deployed but I'll post here when it's live

**Comment by SzymczakJ** [comment]
FYI I'll be OOO until 28.07, @war-in will be taking this issue over from me

**Comment by war-in** [comment]
Hi 👋 AFAIK there is still one thing to do with this issue. @SzymczakJ told me we need to add error handling for the wrong credentials. It's blocked by [this PR](https://github.com/Expensify/App/pull/45251) and should be handled once it's merged. So we're waiting ⌚

**Comment by JmillsExpensify** [comment]
Still waiting on the linked QBO PR above.

**Comment by miljakljajic** [comment]
Is it ready for me to pay? https://github.com/Expensify/App/issues/44669#issuecomment-2232055542

**Comment by hungvu193** [comment]
Noy yet. Still holding for QBO's PR

**Comment by miljakljajic** [comment]
If no payment is needed I'll unassign myself, thank you! Add me back if you need anything from BZ team member.

**Comment by war-in** [comment]
Hey 👋 The [blocking PR](https://github.com/Expensify/App/pull/45251) has been merged and we are ready to start working on this task 🎉 

But there is still one thing to discuss. Are there any designs for the wrong credentials error handling?
In QBO and Xero we show the `Enter credentials` option in the popup but here (and in NetSuite) we could also have an existing connection which we might want to reuse.

@yuwenmemon Could you please help me with getting those designs? 🙏

**Comment by yuwenmemon** [comment]
@war-in Is the question whether we'd want to show the "Reuse Existing Connection" option in addition to "Enter credentials"?

<img width="423" alt="Screenshot 2024-08-21 at 4 31 10 PM" src="https://github.com/user-attachments/assets/3af35fe2-0e64-49f0-854b-6b23add3bc5f">

I'd say No - let's just show Enter credentials and "Disconnect". If one connection is broken, most likely all of them are broken, so reusing existing credentials wouldn't really help in that case.

**Comment by war-in** [comment]
Thanks for the answer! It'll be handled by @SzymczakJ since he'll be back in the office on Monday

**Comment by SzymczakJ** [comment]
I'm back and I'm starting to work on this one!

**Comment by SzymczakJ** [comment]
I've put out the draft PR, but it's gonna work only if we fix one thing on BE.
When giving wrong credentials to Sage Intacct we get a `connections` object with `lasSync` that looks like that:
<img width="552" alt="Screenshot 2024-08-26 at 15 18 26" src="https://github.com/user-attachments/assets/ba6d5354-30f3-4666-ba14-c4d85056fdbf">
we have an `errorMessage` that says "Authentication error" and `isAuthenticationError` set to false. Shouldn't the `isAuthenticationError` be set to true? When we set it to true we can handle this enter credentials flow the same way we do it in Xero and other integrations. @yuwenmemon

**Comment by SzymczakJ** [comment]
Another question: right now all error handling is handled like that:
<img width="924" alt="Screenshot 2024-08-26 at 15 59 13" src="https://github.com/user-attachments/assets/e427b2b2-b035-45b0-90e6-78fbc3321825">
<img width="925" alt="Screenshot 2024-08-26 at 15 59 28" src="https://github.com/user-attachments/assets/e4e35030-84e1-4f00-87ff-e9ca15ac4fb5">
which is not coherent with designs of Sage Intacct
<img width="293" alt="Screenshot 2024-08-26 at 16 03 36" src="https://github.com/user-attachments/assets/a13a8f04-bebe-4028-9bf8-3081534154a0">
I know that there is a design doc about error handling in integrations and that's why I'm wondering, if I should go with the Sage Intacct design doc version of error, or do it the same every other integration does this. Which version is the correct one?

**Comment by yuwenmemon** [comment]
@SzymczakJ I can make sure `isAuthenticationError` is set correctly for you!

For the error message, let's go with the Design Doc version, please. If anything, the other integrations should look more like the Intacct error, but we don't have to worry about fixing that just yet if it's too cumbersome.

**Comment by SzymczakJ** [comment]
Making error match Design Doc version shouldn't be too hard, I'll change it in this PR.

**Comment by SzymczakJ** [comment]
Is BE change already deployed @yuwenmemon?

**Comment by yuwenmemon** [comment]
Yep! Sorry it just went out yesterday

**Comment by yuwenmemon** [comment]
Let me know if you run into any issues

**Comment by SzymczakJ** [comment]
Works as expected, I'll make PR ready for review as a first thing on Monday morning.

**Comment by hungvu193** [comment]
Any update with the PR @SzymczakJ 😄

**Comment by SzymczakJ** [comment]
Sorry, I got ravaged by some other task that came up to be much more troublesome than I thought. I'll work on it at night 😬

**Comment by SzymczakJ** [comment]
Ready for review @hungvu193!

**Comment by hungvu193** [comment]
The PR reached production a while ago, so I think we can close this issue.

@yuwenmemon Can you please help me create a payment issue for this project? Thank you 😄
